File: src/main/java/de/zalando/zally/rules/DefineOAuthScopesRule.java
package de.zalando.zally.rules;

import com.google.common.collect.Sets;
import de.zalando.zally.Violation;
import io.swagger.models.Operation;
import io.swagger.models.Swagger;
import io.swagger.models.auth.OAuth2Definition;
import io.swagger.models.auth.SecuritySchemeDefinition;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static de.zalando.zally.ViolationType.MUST;
import static java.lang.String.format;

public class DefineOAuthScopesRule implements Rule {
    private static final String TITLE = "Define and Assign Access Rights (Scopes)";
    private static final String NOT_DEFINED_DESC = "No valid OAuth2 scope is defined for endpoint %s";
    private static final String INVALID_SCOPE_DESC = "Invalid OAuth2 scope %s for the endpoint %s";
    private static final String RULE_LINK = "http://zalando.github.io/restful-api-guidelines/security/Security.html";
    static final String OAUTH2 = "oauth2";
    static final String UID = "uid";

    @Override
    public List<Violation> validate(Swagger swagger) {
        List<Violation> violations = new ArrayList<>();
        if (swagger == null || swagger.getPaths() == null) {
            return violations;
        }
        Set<String> applicableScopes = new HashSet<>(getDefinedScopes(swagger));
        applicableScopes.add(UID);
        swagger.getPaths().forEach((pathKey, path) -> {
            if (path != null) {
                path.getOperations().forEach(operation -> {
                    violations.addAll(validateOperation(operation, pathKey, applicableScopes));
                });
            }
        });
        return violations;
    }

    // Validate one endpoint
    private List<Violation> validateOperation(Operation operation, String path, Set<String> definedScopes) {
        List<Violation> violations = new ArrayList<>();
        Set<String> actualScopes = extractAppliedScopes(operation);
        Sets.SetView<String> illegalScopes = Sets.difference(actualScopes, definedScopes);
        violations.addAll(illegalScopes
                .stream()
                .map(s -> createViolation(path, format(INVALID_SCOPE_DESC, s, operation.getSummary())))
                .collect(Collectors.toList()));
        if (illegalScopes.size() == actualScopes.size()) {
            violations.add(createViolation(path, format(NOT_DEFINED_DESC, operation.getSummary())));
        }
        return violations;
    }

    // Extract all oauth2 scopes applied to the given operation into a simple list
    private Set<String> extractAppliedScopes(Operation operation) {
        if (operation.getSecurity() == null) {
            return Collections.emptySet();
        }
        return operation.getSecurity()
                .stream()
                .flatMap(map -> map != null ? map.get(OAUTH2).stream() : Stream.empty())
                .filter(Objects::nonNull)
                .collect(Collectors.toSet());
    }

    // create a violation for an operation under a path
    private Violation createViolation(String path, String desc) {
        return new Violation(TITLE, desc, MUST, RULE_LINK, path);
    }

    // get the scopes from security definition
    private Set<String> getDefinedScopes(Swagger swagger) {
        Set<String> result = new HashSet<>();
        if (swagger.getSecurityDefinitions() != null) {
            Map<String, SecuritySchemeDefinition> securityDefinitions = swagger.getSecurityDefinitions();
            SecuritySchemeDefinition securitySchemeDefinition = securityDefinitions.get(OAUTH2);
            if (securitySchemeDefinition != null && securitySchemeDefinition instanceof OAuth2Definition) {
                OAuth2Definition oAuth2Definition = (OAuth2Definition) securitySchemeDefinition;
                result = oAuth2Definition.getScopes().keySet();
            }
        }
        return result;
    }
}

File: server/src/main/java/de/zalando/zally/rule/DefineOAuthScopesRule.kt
package de.zalando.zally.rule

import com.google.common.collect.Sets
import de.zalando.zally.dto.ViolationType.MUST
import io.swagger.models.Operation
import io.swagger.models.Swagger
import io.swagger.models.auth.OAuth2Definition
import org.springframework.stereotype.Component

@Component
class DefineOAuthScopesRule : SwaggerRule() {
    override val title = "Define and Assign Access Rights (Scopes)"
    override val url = "http://zalando.github.io/restful-api-guidelines/security/Security.html" +
        "#must-secure-endpoints-with-oauth-20"
    override val violationType = MUST
    override val code = "M004"
    private val DESC = "Every endpoint must be secured by proper OAuth2 scope"

    override fun validate(swagger: Swagger): Violation? {
        val definedScopes = getDefinedScopes(swagger)
        val hasTopLevelScope = hasTopLevelScope(swagger, definedScopes)
        val paths = swagger.paths.orEmpty().entries.flatMap { (pathKey, path) ->
            path.operationMap.orEmpty().entries.map { (method, operation) ->
                val actualScopes = extractAppliedScopes(operation)
                val undefinedScopes = Sets.difference(actualScopes, definedScopes)
                val unsecured = undefinedScopes.size == actualScopes.size && !hasTopLevelScope
                val msg = when {
                    undefinedScopes.isNotEmpty() ->
                        "undefined scopes: " + undefinedScopes.map { "'${it.second}'" }.joinToString(", ")
                    unsecured ->
                        "no valid OAuth2 scope"
                    else -> null
                }
                if (msg != null) "$pathKey $method has $msg" else null
            }.filterNotNull()
        }
        return if (!paths.isEmpty()) {
            Violation(this, title, DESC, violationType, url, paths)
        } else null
    }

    // get the scopes from security definition
    private fun getDefinedScopes(swagger: Swagger): Set<Pair<String, String>> =
        swagger.securityDefinitions.orEmpty().entries.flatMap { (group, def) ->
            (def as? OAuth2Definition)?.scopes.orEmpty().keys.map { scope -> group to scope }
        }.toSet()

    // Extract all oauth2 scopes applied to the given operation into a simple list
    private fun extractAppliedScopes(operation: Operation): Set<Pair<String, String>> =
        operation.security?.flatMap { groupDefinition ->
            groupDefinition.entries.flatMap { (group, scopes) ->
                scopes.map { group to it }
            }
        }.orEmpty().toSet()

    private fun hasTopLevelScope(swagger: Swagger, definedScopes: Set<Pair<String, String>>): Boolean =
        swagger.security?.any { securityRequirement ->
            securityRequirement.requirements.entries.any { (group, scopes) ->
                scopes.any { scope -> (group to scope) in definedScopes }
            }
        } ?: false

}
